in_calyptia_fleet: Fix inability to start up agent when invalid fleet config fetched

alecholmes opened the pull request:
<!-- Provide summary of changes -->

This fixes a fleet bug where fetching an invalid config then shutting down prevented the agent from starting up again.

**Problem**
The fleet plugin is buggy today in that it will immediately commit newly received config as the current config, implying the new config is valid. This can notably prevent fluent-bit from restarting correctly after receiving an invalid config.

For example:
* Process starts with a valid config
* Fleet plugin receives a new config. It marks it as current and kicks off a hot reload.
* The hot reload fails. The process will use the old working config.
* Process exits
* Process starts
* Process tries to use the newer invalid config instead of the older valid config.

**Change**

This PR changes how configs received by the fleet plugin are committed as valid.

When a new config is received:
* Any existing new (not current or old) config is deleted, since it is being replaced.
* The received config is saved on disk as new and a hot reload is kicked off.
* Subsequent fleet collect callbacks will check if the hot reload of the new config was successful. If so, the new config is promoted to current.

When a process starts up:
* It deletes new config on disk. By virtue of being marked new, that config was not successfully reloaded in the past.
* It loads up using current or old config, if available.
* It then will attempt to fetch any newer config and reload.

<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, #77" -->

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:

- [ ] Example configuration file for the change
- [ ] Debug log output from testing the change
<!--
Please refer to the Developer Guide for instructions on building Fluent Bit with Valgrind support:
https://github.com/fluent/fluent-bit/blob/master/DEVELOPER_GUIDE.md#valgrind
Invoke Fluent Bit and Valgrind as: $ valgrind --leak-check=full ./bin/fluent-bit <args>
-->
- [ ] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

If this is a change to packaging of containers or native binaries then please confirm it works for all targets.

- [ ] Run [local packaging test](./packaging/local-build-all.sh) showing all targets (including any new ones) build.
- [ ] Set `ok-package-test` label to test for all targets (requires maintainer to do).

**Documentation**
<!-- Docs can be edited at https://github.com/fluent/fluent-bit-docs -->
- [ ] Documentation required for this feature

<!--  Doc PR (not required but highly recommended) -->

**Backporting**
<!--
PRs targeting the default master branch will go into the next major release usually.
If this PR should be backported to the current or earlier releases then please submit a PR for that particular branch.
-->
- [ ] Backport to latest stable release.

<!--  Other release PR (not required but highly recommended for quick turnaround) -->
----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

alecholmes committed:
in_calyptia_fleet: Unify fleet config file caching across platforms

Signed-off-by: Alec Holmes <alec@chronosphere.io>

alecholmes committed:
in_calyptia_fleet: Only commit fleet config when successfully reloaded

Signed-off-by: Alec Holmes <alec@chronosphere.io>

zzbamboo committed:
tls: replace NPN with ALPN for client connections

Signed-off-by: Biao Zhu <zhumouren0623@qq.com>

ThomasBrady commented on plugins/in_calyptia_fleet/in_calyptia_fleet.c:
```diff
@@ -2358,15 +2412,27 @@ static int in_calyptia_fleet_init(struct flb_input_instance *in,
+    if (new_config_path != NULL
+        && is_new_fleet_config(ctx, config) == FLB_FALSE) {
+        flb_plg_warn(ctx->ins, "deleting uncommitted new config: %s", new_config_path);
+        if (calyptia_config_delete_by_ref(ctx, "new") == FLB_FALSE) {
```
Should we also rollback old to curr?

ThomasBrady replied:
I guess that's unncessary as we fallback to old in `load_fleet_config()` if curr doesn't exist

ThomasBrady commented on plugins/in_calyptia_fleet/in_calyptia_fleet.c:
```diff
@@ -589,13 +574,15 @@ static void *do_reload(void *data)
     reload->flb->config->conf_path_file = reload->cfg_path;
 
-    flb_free(reload);
     sleep(5);
```
Why do we sleep?

alecholmes replied:
I'm not sure, and this feels like a Chesterton's fence situation. @pwhelan do you have history on this one?

pwhelan replied:
I call sleep just to give the actual reload time to be executed. It might be totally unnecessary.

ThomasBrady commented on plugins/in_calyptia_fleet/in_calyptia_fleet.c:
```diff
@@ -1503,49 +1490,80 @@ static int calyptia_config_delete_by_ref(struct flb_in_calyptia_fleet_config *ct
 
+/**
+* Commits the latest received config as the valid, now-current config.
+* This updates the symlink to the current config file to point to the
```
```suggestion
* This updates the ref file for the current config file to point to the
```
Or similar

ThomasBrady commented on plugins/in_calyptia_fleet/in_calyptia_fleet.c:
```diff
@@ -1503,49 +1490,80 @@ static int calyptia_config_delete_by_ref(struct flb_in_calyptia_fleet_config *ct
-    flb_sds_t cur_ref_filename = fleet_config_ref_filename(ctx, "cur");
-    if (cur_ref_filename != NULL) {
-        unlink(cur_ref_filename);
+    if (exists_fleet_config(ctx, "cur") == FLB_TRUE) {
```
Why this change? shouldn't this be guarded on the ref file existing rather than the referenced config existing?

alecholmes replied:
Agreed. I've updated this function a bit as a result and also tweaked the earlier check around cleaning up uncommitted new config.

dependabot[bot] committed:
workflows: bump actions/checkout from 4 to 5

Bumps [actions/checkout](https://github.com/actions/checkout) from 4 to 5.
- [Release notes](https://github.com/actions/checkout/releases)
- [Changelog](https://github.com/actions/checkout/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/checkout/compare/v4...v5)

---
updated-dependencies:
- dependency-name: actions/checkout
  dependency-version: '5'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

ilove737 committed:
dockerfile: Remove a redundant wget in yum install

Remove a redundant wget in yum install

Signed-off-by: ilove737 <58713761+ilove737@users.noreply.github.com>

alecholmes commented:
PR below this has been merged into OSS so I'm going to close this PR and recreate it in the fluent-bit repo.
